### ISSUE TITLE
Enable using the Parallel.ForEachAsync without a scheduler: invoke the body on the current thread

### ISSUE THREAD

**Issue by theodorzoulias** [dotnet/runtime#82359]
## Background and Motivation

Currently the [`Parallel.ForEachAsync`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.parallel.foreachasync) method cannot be configured to execute the `body` delegate directly on the calling thread. The `body` is always invoked on the specified [`ParallelOptions.TaskScheduler`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.paralleloptions.taskscheduler), which by default is the `TaskScheduler.Default`. Setting this option to `null` does not disable the scheduling. Instead the `body` is scheduled on the `TaskScheduler.Current`, which by default is also the `TaskScheduler.Default`. The only remaining possibility is to pass a custom `TaskScheduler` implementation that invokes synchronously the `Task` instead of scheduling it (like [this one](https://github.com/dotnet/runtime/issues/61318)), but this could brake the API in unexpected ways. [Quote](https://github.com/dotnet/runtime/issues/61318#issuecomment-963353655):

> No. It expects that when it queues work it'll actually be queued, not invoked as part of the queueing operation. There are lots of scheduler variations that can break most asynchronous libraries that expect queueing actually means queueing.

Disabling the scheduling is considered to be beneficial especially in ASP.NET applications. Quoting from the article [Async Programming : Introduction to Async/Await on ASP.NET](https://learn.microsoft.com/en-us/archive/msdn-magazine/2014/october/async-programming-introduction-to-async-await-on-asp-net#asynchronous-code-is-not-a-silver-bullet) by Stephen Cleary: 

> You can kick off some background work by awaiting `Task.Run`, but there’s no point in doing so. In fact, that will actually hurt your scalability by interfering with the ASP.NET thread pool heuristics. If you have CPU-bound work to do on ASP.NET, your best bet is to just execute it directly on the request thread. As a general rule, don’t queue work to the thread pool on ASP.NET.

Stephen Cleary has made many times the distinction between parallelism and asynchronous concurrency, and has advised against the `Parallel.ForEachAsync` because it focuses on the former instead of the later. [Quote](https://stackoverflow.com/questions/70960918/how-can-i-gather-results-from-parallel-async-tasks-in-c/70966172#70966172 "How can I gather results from parallel async tasks in C#?"):

> I disagree. `Parallel.ForEachAsync` is for when you have a complex mixture of CPU-bound and I/O-bound code and need to do both parallelism and asynchronous concurrency. In this case parallelism isn't necessary.

This proposal will allow to use the `Parallel.ForEachAsync` API for achieving pure asynchronous concurrency, without parallelism.

## Proposed API

Redefine the `ParallelOptions.TaskScheduler = null` to mean "no TaskScheduler", when used with the `Parallel.ForEachAsync` API. Only the behavior of the specific configuration is changed. No API modifications are required.

An open question is whether the `SynchronizationContext` should be captured when the `body` is invoked, so that the next invocation of the `body` is on the same context. My preference would be to capture the context.

## Usage Examples

```C#
private async void Button_Click(object sender, EventArgs e)
{
    List<string> urls = GetUrls();
    ParallelOptions options = new()
    {
        MaxDegreeOfParallelism = 10,
        TaskScheduler = null,
    };
    await Parallel.ForEachAsync(urls, options, async (url, ct) =>
    {
        string document = await _httpClient.GetStringAsync(url, ct);
        ListBox1.Items.Add($"{url}: {document.Length} chars");
    });
}
```

In this example the `body` is invoked on the UI thread, allowing the interaction with UI elements. One could think that the same effect could be achieved with the `TaskScheduler.FromCurrentSynchronizationContext()`, but this scheduler has `MaximumConcurrencyLevel = 1`, which interferes with the `MaxDegreeOfParallelism`, effectively [disabling the parallelization](https://github.com/dotnet/runtime/issues/71415).

## Alternative Designs

Maybe adding a new property in the `ParallelOptions` class?

## Risks

It's possible that there is code out there that relies on the existing behavior. I believe that the probabilities are pretty slim though. Even in that case, the fix would be pretty straightforward: replace the `TaskScheduler = null` with `TaskScheduler = TaskScheduler.Current`.

It should be noted that currently the `Parallel.ForEachAsync` API differs from the other `Parallel` APIs on how it interprets the [`MaxDegreeOfParallelism = -1`](https://github.com/dotnet/dotnet-api-docs/pull/7879) configuration (it means `Environment.ProcessorCount` instead of unlimited). So differing also on how it interprets the `TaskScheduler = null` would not be unprecedented.

**Comment by stephentoub** [comment]
This is not something we'll change. ParallelOptions.TaskScheduler being null has a long well-defined, documented meaning, and we're not going to special-case it to mean something else for one specific API. Further, executing the delegate in the calling thread synchronously completely violates the spirit of the API, which is to perform all the work asynchronously from the caller; if a custom scheduler wants to do that, that's its perogative, subject to all the problems that comes with doing so, and we're not going to bake in that behavior, especially not as a default. There's also zero precedent for these APIs capturing the SynchronizationContext to schedule back to. If this all stems from your not liking FromCurrentSynchronizationContext setting MaxConcurrencyLevel to 1, a) there's a separate issue you already opened about that, and b) in that issue there's an 8-line-including-braces implementation of a custom scheduler you can use instead.

Thanks.

> So differing also on how it interprets the TaskScheduler = null would not be unprecedented.

Yes, it would. The API _has_ to treat the max differently, due to the fundamental difference of lack of backpressure from the scheduler. That's not the case for the target scheduler.

**Comment by theodorzoulias** [comment]
> If this all stems from your not liking FromCurrentSynchronizationContext setting MaxConcurrencyLevel to 1, a) there's a separate issue you already opened about that, and b) in that issue there's an 8-line-including-braces implementation of a custom scheduler you can use instead.

No, not at all. I mentioned the [other issue](https://github.com/dotnet/runtime/issues/71415) as a side note. Maybe the usage example that I've chosen is not suitable for demonstrating the spirit of this suggestion. The main goal of this suggestion is to enable asynchronous concurrency without parallelism, for which there is currently no built-in API available. This induces people who endorse this concept to use improvised solutions, usually combinations of a `Task.WhenAll` and a `SemaphoreSlim`, that don't have ideal behavior in case of exceptions. Do you think that the whole concept of *"asynchronous concurrency"* is suspect and shouldn't be promoted, or that it's a valid concept but it should better be offered as a third-party NuGet package? Or that the `Parallel` class is just not the right place to host this functionality?

**Comment by theodorzoulias** [comment]
Btw the intended labels of this issue are [**api-suggestion**](https://github.com/dotnet/runtime/labels/api-suggestion) and [**area-System.Threading.Tasks**](https://github.com/dotnet/runtime/labels/area-System.Threading.Tasks). :-)

**Comment by stephentoub** [comment]
> Do you think that the whole concept of "asynchronous concurrency" is suspect and shouldn't be promoted

There's nothing wrong with wanting to be able to control number of concurrent flows of control separately from parallel execution on threads.  But that's exactly what the other issue is about, and that issue is still open, as if we decide to expose that capability for Parallel.ForEachAsync, doing it as I outlined in https://github.com/dotnet/runtime/issues/71415#issuecomment-1170514595 is how we'd want to do it.

It also _is_ possible with the existing API, again as I outlined in that issue.  The ParallelOptions.MaxDegreeOfParallelism passed to Parallel.ForEachAsync controls the number of concurrent flows of control, and the TaskScheduler itself entirely controls how many delegates are able to run in paralllel.  The TaskScheduler's MaxConcurrencyLevel is meant to reflect that value, as an upper bound.  Today as the other issue notes, ForEachAsync lowers the MaxDegreeOfParallelism to be at most the scheduler's MaxConcurrencyLevel, but that just means if you want to keep them separate, you need to pass in a scheduler that just keeps the default int.MaxValue for MaxConcurrencyLevel.  So for example with:
```C#
internal class Program
{
    static async Task Main()
    {
        int parallel = 0, concurrent = 0;
        _ = Parallel.ForEachAsync(Infinite(), new ParallelOptions { MaxDegreeOfParallelism = 16, TaskScheduler = new WrappedTaskScheduler(new ConcurrentExclusiveSchedulerPair().ExclusiveScheduler) }, async (i, ct) =>
        {
            Interlocked.Increment(ref concurrent);
            if (i % 10 < 8)
            {
                await Task.Delay(1000);
            }
            else
            {
                Interlocked.Increment(ref parallel);
                Thread.Sleep(100);
                Interlocked.Decrement(ref parallel);
            }
            Interlocked.Decrement(ref concurrent);
        });

        while (true)
        {
            Console.WriteLine($"P {parallel}, C {concurrent}");
            Thread.Sleep(100);
        }
    }

    static async IAsyncEnumerable<int> Infinite()
    {
        for (int i = 0; ; i++)
        {
            await Task.Delay(1);
            yield return i;
        }
    }
}

internal sealed class WrappedTaskScheduler : TaskScheduler
{
    private readonly TaskScheduler _scheduler;
    public WrappedTaskScheduler(TaskScheduler scheduler) => _scheduler = scheduler;
    protected override void QueueTask(Task task) => Task.Factory.StartNew(s => TryExecuteTask((Task)s!), task, CancellationToken.None, TaskCreationOptions.DenyChildAttach, _scheduler);
    protected override bool TryExecuteTaskInline(Task task, bool taskWasPreviouslyQueued) => false;
    protected override IEnumerable<Task>? GetScheduledTasks() => null;
}
```
you can see that while the tracked concurrency level frequently tops out at the max of 16, the code execution parallelism is never more than 1, since it's using an exclusive scheduler that only ever runs one thing at a time.

**Comment by theodorzoulias** [comment]
Stephen Toub your approach towards asynchronous concurrency is direct and uncompromising, by enforcing strictly the "no parallelization" rule. But I think that Stephen Cleary has something else in mind. Please take a look at a [recent implementation](https://stackoverflow.com/questions/75024353/parallel-foreachasync-keep-sort-order/75033114#75033114) that he shared:

> _Asynchronous concurrency uses `Task.WhenAll` to get the results of multiple asynchronous operations. It can also use `SemaphoreSlim` for throttling. Something like this:_
>
> ```C#
> var mutex = new SemaphoreSlim(20);
> var results = await Task.WhenAll(model.DestinationFiles.Select(async file =>
> {
>     await mutex.WaitAsync();
>     try
>     {
>         var storeAsync = await _fileServerService.Init(displayUrl).StoreAsync(
>             file.FileInfo, false, file.OutputFileName);
>         return new ConverterConvertResultFile(storeAsync);
>     }
>     finally { mutex.Release(); }
> });
> convertResultDto.Files.AddRange(results);
> ```

Notice that there is nothing preventing the parallel invocation of the `StoreAsync`. Initially the `StoreAsync` is invoked synchronously 20 times, until the `mutex` reaches its maximum capacity. Then the next `StoreAsync` will be invoked on whatever thread a previous `StoreAsync` was completed. Assuming that a `SynchronizationContext` is not present, multiple subsequent `StoreAsync` invocations could potentially occur in parallel. I think that the emphasis is on avoiding any explicit attempt to control where the asynchronous operations are invoked. Just let the whole operation unfold naturally without any external intervention, and if some accidental parallelization occurs, then so be it. Assuming that the `ConverterConvertResultFile` is a lightweight CPU-bound operation that is dwarfed by the in-flight duration of the I/O-bound `StoreAsync`, the amount of parallelization that is expected to happen should be minimal anyway.

My suggestion, about redefining the meaning of `ParallelOptions.TaskScheduler = null`, aims at making the above ASP.NET-friendly approach possible, using a robust, tested and well behaving API instead of `Task.WhenAll`+`SemaphoreSlim` improvisations. In this specific case there would still be the issue of how to collect the `var results`, but this is probably an easier problem to solve compared to fixing the problematic behavior in case of exceptions. And in general "polluting" application-level code with `SemaphoreSlim`s and other low-level synchronization primitives is not great. This kind of functionality should better be abstracted IMHO.

**Comment by stephentoub** [comment]
> Notice that there is nothing preventing the parallel invocation of the StoreAsync.

There isn't in my example, either. Just change the:
```C#
new ConcurrentExclusiveSchedulerPair().ExclusiveScheduler
```
to
```C#
new ConcurrentExclusiveSchedulerPair(10).ConcurrentScheduler
```
The scheduling decisions of how much is able to run in parallel, where, and how are entirely up to the scheduler, separate from how many concurrent flows of control are involved.

**Comment by theodorzoulias** [comment]
@stephentoub regarding using the `ExclusiveScheduler`/`ConcurrentScheduler` of the `ConcurrentExclusiveSchedulerPair`, wrapped in a `WrappedTaskScheduler`, for limiting the CPU-bound parallelization without affecting the I/O-bound asynchronous concurrency, my understanding is that it will cause scheduling of each and every `await` operation. Not only each `body(item, ct)` operation will be scheduled, but also each individual `await` inside the `body`. Which goes against the advice offered in the [aforementioned article](https://learn.microsoft.com/en-us/archive/msdn-magazine/2014/october/async-programming-introduction-to-async-await-on-asp-net#asynchronous-code-is-not-a-silver-bullet) about async programming on ASP.NET:

> As a general rule, don’t queue work to the thread pool on ASP.NET.

AFAIK the [`ConcurrentExclusiveSchedulerPair`](https://learn.microsoft.com/en-us/dotnet/api/system.threading.tasks.concurrentexclusiveschedulerpair) by default wraps the `TaskScheduler.Default`, which schedules the queued tasks to the `ThreadPool`.

Is it possible that the advice given in that article has become obsolete, because the `ThreadPool`'s heuristics have been improved since the publication of that article (October 2014)?

**Comment by stephentoub** [comment]
> Assuming that the concept of asynchronous concurrency is valid, and there is value at offering this functionality out of the box, would you consider hosting it somewhere else, in a new Concurrent class for example?

No, I don't believe it's worth it.
